### PR TITLE
Add checked operation methods to Duration

### PR DESCRIPTION
**Pull request by eugene-bulkin**
Addresses #35774.

### CODE REVIEW

**Comment by sfackler** [comment]
Can we update the `Add`, `Sub` etc impls to call into these methods to avoid code duplication?

File: src/libstd/time/duration.rs [hunk 1]
```diff
@@ -97,6 +97,130 @@ impl Duration {
...
+                }
+            };
+            debug_assert!(nanos < NANOS_PER_SEC);
+            Some(Duration { secs: secs, nanos: nanos })
+        } else {
+            None
+        }
+    }
+
+    /// Checked integer multiplication. Computes `self * other`, returning
```
Comment (nagisa) [inline src/libstd/time/duration.rs]: > Checked duration multiplication

**Comment by eugene-bulkin** [comment]
@sfackler What's the best way to do that? Only catch I can see is `checked_op` can fail in a variety of different ways but returns `None` every time, so there's no way to know exactly *why* it failed, while the trait impls will panic with the actual error reason.

**Comment by sfackler** [comment]
It looks like they only fail due to overflow, so you should just be able to do something like

```rust
fn add(self, other: Duration) -> Duration {
    self.checked_add(other).expect("overflow when adding durations")
}
```

File: src/libstd/time/duration.rs (hunk 1 above)
Comment (eugene-bulkin) [inline src/libstd/time/duration.rs]: Fixed, thanks!

**Comment by eugene-bulkin** [comment]
So how about this:

- `add` has the expect message "overflow when adding durations"
- `sub` has the expect message "overflow when subtracting durations"
- `mul` has the expect message "overflow when multiplying duration by scalar"
- `div` has the expect message "divide by zero error when dividing duration by scalar"

Are those reasonable? Or are the last two too wordy?

**Comment by sfackler** [comment]
Those all seem reasonable. I wouldn't worry too much about the wordyness - people hopefully shouldn't be seeing them very often :)

**Comment by eugene-bulkin** [comment]
Looks like doc-tests are failing. I'll get those fixed ASAP.

**Comment by alexcrichton** [comment]
Looks good to me! r=me when the tests are passing

**Comment by eugene-bulkin** [comment]
Doc-tests failing now because I marked the functions as unstable.

Should I leave the marked as unstable and add the `#![feature(duration_checked_ops)]` part in the attributes? Or should I mark them as stable?

**Comment by sfackler** [comment]
You should add the attributes in the examples.

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!
